File: python/examples/asr_file_demo.py
#!/usr/bin/env python3
"""
ASR 静态文件识别示例

只依赖 spacemit_asr 模块，用于识别音频文件。

Usage:
    python asr_file_demo.py audio.wav
    python asr_file_demo.py --list-backends
"""

import sys
import os
import argparse


def main():
    parser = argparse.ArgumentParser(description='ASR 静态文件识别')
    parser.add_argument('file', nargs='?', help='要识别的音频文件（例如 WAV 或 MP3）')
    parser.add_argument('--engine', default='sensevoice',
                        choices=['sensevoice', 'zipformer', 'funasr', 'qwen3-asr',
                                 'gemma4-asr'],
                        help='识别引擎 (默认: sensevoice)')
    parser.add_argument('--model-dir', '-m', default=None,
                        help='本地模型目录 (SenseVoice/Zipformer 可选)')
    parser.add_argument('--language', '-l', default='auto',
                        choices=['zh', 'en', 'ja', 'ko', 'yue', 'auto'],
                        help='识别语言 (默认: auto)')
    parser.add_argument('--list-backends', action='store_true',
                        help='列出可用后端')
    parser.add_argument('--provider', '-p', default='spacemit',
                        choices=['cpu', 'spacemit'],
                        help='执行提供程序 (默认: spacemit)')
    parser.add_argument('--rounds', '-r', type=int, default=1,
                        help='重复识别轮次 (默认: 1)')
    parser.add_argument('--enable-emotion', action='store_true',
                        help='启用 SenseVoice 情绪识别 (默认关闭)')
    parser.add_argument('--endpoint', default=None,
                        help='llama-server URL')
    parser.add_argument('--model', default=None,
                        help='llama-server model tag')
    parser.add_argument('--timeout', type=int, default=60,
                        help='llama-server 请求超时秒数 (默认: 60)')
    parser.add_argument('--task', default='transcribe',
                        choices=['transcribe', 'translate'],
                        help='Gemma4 任务 (默认: transcribe)')

    args = parser.parse_args()

    # 导入 spacemit_asr
    try:
        import spacemit_asr
    except ImportError as e:
        print("错误: 无法导入 spacemit_asr 模块")
        print("请确保已设置 PYTHONPATH:")
        print("  export PYTHONPATH=/path/to/asr/build/python:/path/to/asr/python")
        print(f"\n详细错误: {e}")
        sys.exit(1)

    print(f"ASR 版本: {spacemit_asr.__version__}")

    if args.list_backends:
        backends = ", ".join(
            str(backend) for backend in spacemit_asr.Engine.get_available_backends())
        print(f"可用后端: {backends}")
        return

    if not args.file:
        parser.print_help()
        return

    if not os.path.exists(args.file):
        print(f"错误: 文件不存在 - {args.file}")
        sys.exit(1)

    # 配置
    lang_map = {
        'zh': spacemit_asr.Language.ZH,
        'en': spacemit_asr.Language.EN,
        'ja': spacemit_asr.Language.JA,
        'ko': spacemit_asr.Language.KO,
        'yue': spacemit_asr.Language.YUE,
        'auto': spacemit_asr.Language.AUTO,
    }

    if args.engine == 'sensevoice':
        config = spacemit_asr.Config.sensevoice(
            args.model_dir or '~/.cache/models/asr/sensevoice',
            enable_emotion=args.enable_emotion)
    elif args.engine == 'zipformer':
        config = spacemit_asr.Config.zipformer(
            args.model_dir or '~/.cache/models/asr/zipformer')
    elif args.engine == 'funasr':
        config = spacemit_asr.Config.funasr(
            endpoint=args.endpoint or
            'http://127.0.0.1:8063/v1/audio/transcriptions',
            model=args.model or 'funasr',
            timeout=args.timeout)
    elif args.engine == 'qwen3-asr':
        config = spacemit_asr.Config.qwen3_asr(
            endpoint=args.endpoint or
            'http://127.0.0.1:8063/v1/chat/completions',
            model=args.model or 'qwen3-asr',
            timeout=args.timeout)
    else:
        config = spacemit_asr.Config.gemma4_asr(
            endpoint=args.endpoint or
            'http://127.0.0.1:8063/v1/audio/transcriptions',
            model=args.model or 'gemma4-asr',
            timeout=args.timeout,
            task=args.task)

    config.language = lang_map[args.language]
    config.punctuation_enabled = True
    config.provider = args.provider
    if args.engine == 'sensevoice':
        config.enable_emotion = args.enable_emotion

    # 识别
    print(f"\n文件: {args.file}")
    print(f"引擎: {args.engine}")
    print(f"语言: {args.language}")
    if args.engine == 'gemma4-asr':
        print(f"任务: {args.task}")
    print(f"Provider: {args.provider}")
    print(f"情绪识别: {'启用' if args.enable_emotion else '禁用'}")
    print("正在识别...\n")

    with spacemit_asr.Engine(config) as engine:
        import time
        import numpy as np
        if args.engine not in ('funasr', 'qwen3-asr', 'gemma4-asr'):
            print(">>> Warmup...")
            silence = np.zeros(8000, dtype=np.float32)
            t0 = time.monotonic()
            engine.recognize(silence)
            warmup_ms = (time.monotonic() - t0) * 1000
            print(f"Warmup done: {warmup_ms:.0f} ms\n")

        total_audio_ms = 0
        total_proc_ms = 0
        for round_idx in range(args.rounds):
            result = engine.recognize_file(args.file)
            total_audio_ms += result.audio_duration_ms
            total_proc_ms += result.processing_time_ms
            print(f"[轮次 {round_idx + 1}/{args.rounds}] {result.text}")
            if result.emotion:
                print(f"情绪: {result.emotion}")

        if args.rounds > 1:
            print(f"\n--- 汇总 ({args.rounds} 轮) ---")
            print(f"平均音频时长: {total_audio_ms // args.rounds} ms")
            print(f"平均处理时间: {total_proc_ms // args.rounds} ms")
            print(f"平均 RTF: {total_proc_ms / max(total_audio_ms, 1):.3f}")
        else:
            print(f"音频时长: {result.audio_duration_ms} ms")
            print(f"处理时间: {result.processing_time_ms} ms")
            print(f"RTF: {result.rtf:.3f}")


if __name__ == "__main__":
    main()

File: examples/asr_file_demo.cpp
/*
 * Copyright (C) 2026 SpacemiT (Hangzhou) Technology Co. Ltd.
 * SPDX-License-Identifier: Apache-2.0
 */

/**
 * SpacemitAudioSDK 静态文件识别示例
 *
 * Usage:
 *   ./asr_file_demo <audio1.wav> [audio2.wav ...] [--model-dir DIR] [--rounds N] [--provider EP] [--enable-emotion]
 *
 * Examples:
 *   ./asr_file_demo ~/test.wav
 *   ./asr_file_demo a.wav b.wav c.wav
 *   ./asr_file_demo a.wav b.wav --model-dir ~/.cache/models/asr/sensevoice
 */

#include <chrono>
#include <cstdlib>
#include <iomanip>
#include <iostream>
#include <memory>
#include <string>
#include <vector>

#include "asr_service.h"

struct FileResult {
    int round;
    std::string file;
    double audio_ms;
    double process_ms;
    double rtf;
    std::string text;
    std::string emotion;
};

class ErrorCaptureCallback : public SpacemiT::AsrEngineCallback {
public:
    void Clear() {
        last_error_.clear();
    }

    const std::string& LastError() const {
        return last_error_;
    }

    void OnError(std::shared_ptr<SpacemiT::RecognitionResult> result) override {
        if (result && !result->GetText().empty()) {
            last_error_ = result->GetText();
        } else {
            last_error_ = "unknown ASR error";
        }
    }

private:
    std::string last_error_;
};

void printUsage(const char* program) {
    std::cout << "Usage: " << program
        << " <audio1> [audio2 ...] [OPTIONS]"
        << std::endl;
    std::cout << std::endl;
    std::cout << "Arguments:" << std::endl;
    std::cout << "  audio files   One or more audio files (for example WAV or MP3)" << std::endl;
    std::cout << "  --engine      Engine: sensevoice | zipformer | funasr | qwen3-asr | gemma4-asr"
        << " (default: sensevoice)" << std::endl;
    std::cout << "  --model-dir   Path to SenseVoice model directory" << std::endl;
    std::cout << "                Default: ~/.cache/models/asr/sensevoice" << std::endl;
    std::cout << "  --rounds N    Run N rounds of recognition (default: 1)" << std::endl;
    std::cout << "  --provider    EP: cpu | spacemit (default: spacemit)" << std::endl;
    std::cout << "  --hotwords    Comma-separated hotwords (e.g. \"SpacemiT,进迭时空\")" << std::endl;
    std::cout << "  --hotword-boost  Hotword boost weight (default: 2.0)" << std::endl;
    std::cout << "  --enable-emotion  Enable SenseVoice emotion recognition (default: off)" << std::endl;
    std::cout << "  --endpoint    llama-server URL" << std::endl;
    std::cout << "  --model       llama-server model tag" << std::endl;
    std::cout << "  --timeout     llama-server timeout in seconds (default: 60)" << std::endl;
    std::cout << "  --task        Gemma4 task: transcribe | translate"
        << " (default: transcribe)" << std::endl;
    std::cout << std::endl;
    std::cout << "Examples:" << std::endl;
    std::cout << "  " << program << " ~/test.wav" << std::endl;
    std::cout << "  " << program << " a.wav --hotwords \"SpacemiT,进迭时空\" --hotword-boost 3.0" << std::endl;
    std::cout << "  " << program
                << " a.wav b.wav --engine qwen3-asr"
                << " --endpoint http://asr-server.example.com:8063/v1/chat/completions"
                << std::endl;
    std::cout << "  " << program
                << " a.wav --engine funasr"
                << " --endpoint http://asr-server.example.com:8063/v1/audio/transcriptions"
                << std::endl;
    std::cout << "  " << program
                << " 024_ja_funasr_sample.mp3 --engine gemma4-asr --task translate"
                << std::endl;
    std::cout << "  " << program << " a.wav b.wav --model-dir /path/to/models" << std::endl;
}

std::string expandHome(const std::string& path) {
    if (!path.empty() && path[0] == '~') {
        const char* home = getenv("HOME");
        if (home) return std::string(home) + path.substr(1);
    }
    return path;
}

int main(int argc, char* argv[]) {
    if (argc < 2 || std::string(argv[1]) == "-h" || std::string(argv[1]) == "--help") {
        printUsage(argv[0]);
        return (argc < 2) ? 1 : 0;
    }

    // Parse args
    std::vector<std::string> audio_files;
    std::string engine_name = "sensevoice";
    std::string model_dir;
    bool model_dir_set = false;
    std::string provider = "spacemit";
    std::string hotwords_str;
    float hotword_boost = 2.0f;
    std::string endpoint;
    std::string model_tag;
    bool endpoint_set = false;
    bool model_tag_set = false;
    int timeout = 60;
    int rounds = 1;
    bool enable_emotion = false;
    std::string task = "transcribe";

    for (int i = 1; i < argc; i++) {
        std::string arg = argv[i];
        if (arg == "--engine" && i + 1 < argc) {
            engine_name = argv[++i];
        } else if (arg == "--model-dir" && i + 1 < argc) {
            model_dir = argv[++i];
            model_dir_set = true;
        } else if (arg == "--rounds" && i + 1 < argc) {
            rounds = std::atoi(argv[++i]);
            if (rounds < 1) rounds = 1;
        } else if (arg == "--provider" && i + 1 < argc) {
            provider = argv[++i];
        } else if (arg == "--hotwords" && i + 1 < argc) {
            hotwords_str = argv[++i];
        } else if (arg == "--hotword-boost" && i + 1 < argc) {
            hotword_boost = std::stof(argv[++i]);
        } else if (arg == "--enable-emotion") {
            enable_emotion = true;
        } else if (arg == "--endpoint" && i + 1 < argc) {
            endpoint = argv[++i];
            endpoint_set = true;
        } else if (arg == "--model" && i + 1 < argc) {
            model_tag = argv[++i];
            model_tag_set = true;
        } else if (arg == "--timeout" && i + 1 < argc) {
            timeout = std::atoi(argv[++i]);
        } else if (arg == "--task" && i + 1 < argc) {
            task = argv[++i];
        } else {
            audio_files.push_back(expandHome(argv[i]));
        }
    }

    if (audio_files.empty()) {
        std::cerr << "Error: no audio files specified" << std::endl;
        return 1;
    }
    if (engine_name == "gemma4-asr" &&
        task != "transcribe" &&
        task != "translate") {
        std::cerr << "Error: --task must be transcribe or translate" << std::endl;
        return 1;
    }

    // Initialize engine once
    std::cout << "========================================" << std::endl;
    std::cout << "    SpacemitAudioSDK 文件识别测试" << std::endl;
    std::cout << "========================================" << std::endl;
    std::cout << std::endl;

    std::cout << ">>> 创建 ASR 引擎 (" << engine_name << ")..." << std::endl;
    SpacemiT::AsrConfig config = SpacemiT::AsrConfig::Preset(engine_name);
    config.language = "auto";
    config.punctuation = true;
    config.enable_emotion = enable_emotion;

    const bool is_server_backend =
        engine_name == "funasr" ||
        engine_name == "qwen3-asr" ||
        engine_name == "gemma4-asr";
    if (is_server_backend) {
        if (endpoint_set) config.endpoint = endpoint;
        if (model_tag_set) config.model = model_tag;
        config.timeout = timeout;
        if (engine_name == "gemma4-asr") {
            config.task = task;
        }
    } else {
        if (model_dir_set) {
            config.model_dir = expandHome(model_dir);
        }
        config.provider = provider;
    }

    // 解析热词
    if (!hotwords_str.empty()) {
        std::vector<std::string> words;
        std::string word;
        for (char c : hotwords_str) {
            if (c == ',') {
                if (!word.empty()) {
                    words.push_back(word);
                    word.clear();
                }
            } else {
                word += c;
            }
        }
        if (!word.empty()) words.push_back(word);
        config.hotwords = words;
        config.hotword_boost = hotword_boost;
    }

    auto engine = std::make_shared<SpacemiT::AsrEngine>(config);
    if (!engine->IsInitialized()) {
        std::cerr << "引擎初始化失败!" << std::endl;
        return 1;
    }

    auto error_callback = std::make_shared<ErrorCaptureCallback>();
    engine->SetCallback(error_callback);

    auto cfg = engine->GetConfig();
    std::cout << "引擎类型: " << cfg.engine << std::endl;
    std::cout << "语言: " << cfg.language << std::endl;
    std::cout << "标点: " << (cfg.punctuation ? "启用" : "禁用") << std::endl;
    std::cout << "情绪识别: " << (cfg.enable_emotion ? "启用" : "禁用") << std::endl;
    std::cout << "采样率: " << cfg.sample_rate << " Hz" << std::endl;
    if (is_server_backend) {
        std::cout << "Endpoint: " << cfg.endpoint << std::endl;
        std::cout << "Model: " << cfg.model << std::endl;
        std::cout << "Timeout: " << cfg.timeout << "s" << std::endl;
        if (engine_name == "gemma4-asr") {
            std::cout << "Task: " << cfg.task << std::endl;
        }
    } else {
        std::cout << "Provider: " << provider << std::endl;
    }
    if (!cfg.hotwords.empty()) {
        std::cout << "热词: ";
        for (size_t j = 0; j < cfg.hotwords.size(); ++j) {
            if (j > 0) std::cout << ", ";
            std::cout << cfg.hotwords[j];
        }
        std::cout << " (boost=" << cfg.hotword_boost << ")" << std::endl;
    }
    std::cout << "文件数: " << audio_files.size() << std::endl;
    std::cout << "轮次: " << rounds << std::endl;
    std::cout << std::endl;

    // --- Warmup: 跑一次哑推理，加热 EP JIT 缓存 ---
    if (!is_server_backend) {
        std::cout << ">>> Warmup (excluded from benchmark)..." << std::endl;
        {
            std::vector<float> silence(8000, 0.0f);  // 0.5s 静音 @16kHz
            auto t0 = std::chrono::steady_clock::now();
            engine->Recognize(silence, 16000);
            auto t1 = std::chrono::steady_clock::now();
            double warmup_ms = std::chrono::duration<double, std::milli>(t1 - t0).count();
            std::cout << "Warmup done: " << std::fixed << std::setprecision(0)
                << warmup_ms << " ms" << std::endl;
        }
        std::cout << std::endl;
    }

    // Recognize each file, multiple rounds
    std::vector<FileResult> results;
    bool had_failure = false;

    for (int round = 0; round < rounds; ++round) {
        if (rounds > 1) {
            std::cout << "======== 第 " << (round + 1) << "/" << rounds << " 轮 ========" << std::endl;
        }

        for (size_t i = 0; i < audio_files.size(); ++i) {
            const auto& file = audio_files[i];
            std::cout << "----------------------------------------" << std::endl;
            std::cout << "[" << (i + 1) << "/" << audio_files.size() << "] " << file << std::endl;

            error_callback->Clear();
            auto result = engine->Call(file);

            if (result && !result->IsEmpty()) {
                FileResult fr;
                fr.round = round + 1;
                fr.file = file;
                fr.text = result->GetText();
                fr.emotion = result->GetEmotion();
                fr.audio_ms = result->GetAudioDuration();
                fr.process_ms = result->GetProcessingTime();
                fr.rtf = result->GetRTF();
                results.push_back(fr);

                std::cout << "文本: " << fr.text << std::endl;
                if (!fr.emotion.empty()) {
                    std::cout << "情绪: " << fr.emotion << std::endl;
                }
                std::cout << "音频: " << std::fixed << std::setprecision(0) << fr.audio_ms << " ms"
                    << "  处理: " << fr.process_ms << " ms"
                    << "  RTF: " << std::setprecision(3) << fr.rtf << std::endl;
            } else if (!error_callback->LastError().empty()) {
                had_failure = true;
                std::cerr << "识别失败: " << error_callback->LastError() << std::endl;
            } else {
                had_failure = true;
                std::cerr << "识别失败或未检测到语音" << std::endl;
            }
            std::cout << std::endl;
        }
    }

    // Summary table
    if (results.size() > 1) {
        std::cout << "========================================" << std::endl;
        std::cout << "              汇总" << std::endl;
        std::cout << "========================================" << std::endl;
        std::cout << std::left << std::setw(40) << "文件"
            << std::right << std::setw(10) << "音频(ms)"
            << std::setw(10) << "处理(ms)"
            << std::setw(8) << "RTF" << std::endl;
        std::cout << std::string(68, '-') << std::endl;

        double total_audio = 0, total_process = 0;
        for (const auto& r : results) {
            // Extract filename from path
            std::string name = r.file;
            size_t pos = name.rfind('/');
            if (pos != std::string::npos) name = name.substr(pos + 1);

            std::cout << std::left << std::setw(40) << name
                << std::right << std::fixed
                << std::setw(10) << std::setprecision(0) << r.audio_ms
                << std::setw(10) << r.process_ms
                << std::setw(8) << std::setprecision(3) << r.rtf << std::endl;
            total_audio += r.audio_ms;
            total_process += r.process_ms;
        }

        std::cout << std::string(68, '-') << std::endl;
        std::cout << std::left << std::setw(40) << "Total"
            << std::right << std::fixed
            << std::setw(10) << std::setprecision(0) << total_audio
            << std::setw(10) << total_process
            << std::setw(8) << std::setprecision(3) << (total_process / total_audio) << std::endl;
    }

    engine.reset();
    std::cout << std::endl << "Done." << std::endl;
    _Exit(had_failure ? 1 : 0);
}
